### PR TITLE
Update Error prone configuration; Fix bugs

### PR DESCRIPTION
**Pull request by leventov**
Fixed three bugs in production code:
 - In `SketchEstimateWithErrorBounds`, `datasketches` module
 - In `GranularUnprocessedPathSpec`, `indexing-hadoop` module
 - In `Log4JShutdownPropertyChecker`, `services` module

+ More bugs in test code.

@drcrallen [commented](https://github.com/druid-io/druid/pull/1271#issuecomment-93563584) in 2015:

> Since there is some uncertainty about the behavior of the compiler vs what is really being used currently, and if it modifies byte code at all, I have modified the PR to have a separate POM which contains the error checking stuff.

<s>[Error Prone is a javac plugin, for a version of javac that corresponds roughly to JDK 9 EA Build 159](https://github.com/google/error-prone/issues/620#issuecomment-299320588). It's bytecode output shouldn't be more risky to use than the default `javac` output. Maybe even safer, because JDK 9's javac [fixes some bugs from JDK 8's javac](https://groups.google.com/d/msg/error-prone-discuss/B2aOPzx11lo/ZEdmkR5ZAgAJ).</s> The stability and safety of this compiler is yet questionable, according to https://github.com/druid-io/druid/pull/4252#issuecomment-300046036. However it turned out unnecessary to make Error prone the default compiler; It's configuration was just updated, keeping it in an optional profile, enabled in travis via `mvn -Pstrict compile test-compile`.

### CODE REVIEW

File: indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java [hunk 1]
```diff
@@ -74,13 +75,11 @@ public Job addInputPaths(HadoopDruidIndexerConfig config, Job job) throws IOExce
     final FileSystem fs = betaInput.getFileSystem(job.getConfiguration());
     final Granularity segmentGranularity = config.getGranularitySpec().getSegmentGranularity();
 
-    Map<Long, Long> inputModifiedTimes = new TreeMap<>(
-        Comparators.inverse(Comparators.comparable())
-    );
+    Map<Long, Long> inputModifiedTimes = new TreeMap<>(Ordering.natural());
```
Comment (gianm) [inline indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java]: Should it still be reversed?

File: server/src/test/java/io/druid/client/client/BatchServerInventoryViewTest.java [hunk 2]
```diff
@@ -290,7 +290,9 @@ public void testRunWithFilterCallback() throws Exception
       @Override
       public int compare(DataSegment o1, DataSegment o2)
       {
-        return o1.getInterval().equals(o2.getInterval()) ? 0 : -1;
+        Interval i1 = o1.getInterval();
+        Interval i2 = o2.getInterval();
+        return Long.compare(i1.getStartMillis(), i2.getStartMillis());
```
Comment (gianm) [inline server/src/test/java/io/druid/client/client/BatchServerInventoryViewTest.java]: How about `Comparators.intervalsByStartThenEnd()`?

File: indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java (hunk 1 above)
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java]: either way it should be covered by unit tests if it is not.

**Comment by gianm** [comment]
Antlr-generated code is triggering a lot of warnings, is there a way to ignore that?

There's also a test failure in DruidCoordinatorBalancerTest that might be related to the changes.

**Comment by leventov** [comment]
@gianm yes, in a module pom (druid-common for Antlr) compiler plugin configuration could be updated and certain bug patterns turned off. I just did that, because I'm going to make MissingOverride a error rather than warning in a follow-up PR. Also comes to Avro and Thrift.

File: indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java (hunk 1 above)
Comment (leventov) [inline indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java]: Reverted.

@drcrallen this order is used only in the body of the method, later the contents of this collection are converted and moved to another collection with different ordering. So I doubt testing is needed. Also I don't know *why* `inputModifiedTimes` should have reversed ordering, it dates back to "Initial commit of Druid OSS" in 2012 by @cheddar.

File: server/src/test/java/io/druid/client/client/BatchServerInventoryViewTest.java (hunk 2 above)
Comment (leventov) [inline server/src/test/java/io/druid/client/client/BatchServerInventoryViewTest.java]: Replaced, thanks

**Comment by leventov** [comment]
@gianm didn't add suppressions, because it's quite contrary to the point of compiler warnings to "annoy" somebody and make him finally to fix problems. Also a large portion of all warnings are [MissingOverride] that I'm going to make an error pattern and resolve in a follow-up PR.

**Comment by drcrallen** [comment]
This was brought up previously as a default compiler in https://github.com/druid-io/druid/pull/1271 and it was noted that unless there's a very strong way to track errorprone vs the default java compiler, it is not worth the complication of using a non-default compiler as the binary creator. Which is why the profile for `strict` was added. Can you show the guarantees that errorprone and the default maven compiler produce the exact same bytecode?

**Comment by SerialVelocity** [comment]
Just as a note, we think we've found a bug in the JDK9 compiler that is breaking things for us. It may be worth waiting until JDK9 is stable.

**Comment by leventov** [comment]
@gianm @drcrallen ok, I finally reverted `pom.xml` almost to the same state as before this PR, so Error Prone is optional, enabled via `mvn -Pstrict` in Travis. I updated `travis.yml` to omit `mvn install -DskipTests` step, because it was causing build timeouts, and seems unnecessary since we already do `mvn test` and `mvn -Pstrict compile test-compile`.

**Comment by leventov** [comment]
@SerialVelocity interesting. Could you please share any details? E. g. link to OpenJDK mailing list thread, where this bug is discussed?

File: processing/src/main/java/io/druid/query/QueryContexts.java [hunk 3]
```diff
@@ -111,7 +111,7 @@
   public static <T> long getTimeout(Query<T> query, long defaultTimeout)
   {
     final long timeout = parseLong(query, TIMEOUT_KEY, defaultTimeout);
-    Preconditions.checkState(timeout >= 0, "Timeout must be a non negative value, but was [%d]", timeout);
+    Preconditions.checkState(timeout >= 0, "Timeout must be a non negative value, but was [%s]", timeout);
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/QueryContexts.java]: What's wrong with `%d`?

File: processing/src/main/java/io/druid/segment/MetricHolder.java [hunk 4]
```diff
@@ -113,6 +113,9 @@ public static void writeToChannel(MetricHolder holder, WritableByteChannel out)
       case FLOAT:
         holder.floatType.writeToChannel(out);
         break;
+      case LONG:
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/MetricHolder.java]: I don't think this method is used at all. Let's just delete it.

File: server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java [hunk 5]
```diff
@@ -431,7 +432,7 @@ public void childEvent(CuratorFramework client, PathChildrenCacheEvent event) th
 
             break;
           }
-          case INITIALIZED:
+          case INITIALIZED: {
```
Comment (gianm) [inline server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java]: What's the purpose of adding braces here?

File: processing/src/main/java/io/druid/query/QueryContexts.java (hunk 3 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/QueryContexts.java]: Surprisingly it's not supported by Preconditions. Check https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Preconditions.html

File: processing/src/main/java/io/druid/segment/MetricHolder.java (hunk 4 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/MetricHolder.java]: Removed

File: server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java (hunk 5 above)
Comment (leventov) [inline server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java]: For consistency, other cases in this switch have such braces

File: processing/src/main/java/io/druid/query/QueryContexts.java (hunk 3 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/QueryContexts.java]: Oh, that is surprising!

File: server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java (hunk 5 above)
Comment (gianm) [inline server/src/main/java/io/druid/curator/inventory/CuratorInventoryManager.java]: Got it.

File: indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java [hunk 6]
```diff
@@ -74,13 +75,11 @@ public Job addInputPaths(HadoopDruidIndexerConfig config, Job job) throws IOExce
...
 
-    Map<Long, Long> inputModifiedTimes = new TreeMap<>(
-        Comparators.inverse(Comparators.comparable())
-    );
+    Map<Long, Long> inputModifiedTimes = new TreeMap<>(Ordering.natural().reverse());
 
     for (FileStatus status : FSSpideringIterator.spiderIterable(fs, betaInput)) {
       final DateTime key = segmentGranularity.toDate(status.getPath().toString());
-      final Long currVal = inputModifiedTimes.get(key);
+      final Long currVal = inputModifiedTimes.get(key.getMillis());
```
Comment (drcrallen) [inline indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java]: Was this a bug?

File: server/src/test/java/io/druid/server/coordinator/DruidCoordinatorBalancerTester.java [hunk 7]
```diff
@@ -46,7 +46,6 @@ protected void moveSegment(
 
     if (!toPeon.getSegmentsToLoad().contains(segmentToMove) &&
         !currentlyMovingSegments.get("normal").containsKey(segmentName) &&
-        !toServer.getSegments().containsKey(segmentName) &&
```
Comment (drcrallen) [inline server/src/test/java/io/druid/server/coordinator/DruidCoordinatorBalancerTester.java]: Why is this removed?

File: indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java (hunk 6 above)
Comment (leventov) [inline indexing-hadoop/src/main/java/io/druid/indexer/path/GranularUnprocessedPathSpec.java]: Yes, this Map has Long, not DateTime keys

File: server/src/test/java/io/druid/server/coordinator/DruidCoordinatorBalancerTester.java (hunk 7 above)
Comment (leventov) [inline server/src/test/java/io/druid/server/coordinator/DruidCoordinatorBalancerTester.java]: Because toServer's segments are not updated in the block of this if statement (unlike in production code, from where this condition was apparently copied), and this expression is either always true or false. It wasn't noticed before because due to the other bug, fixed a few lines below, all tests using `DruidCoordinatorBalancerTester` basically always passed.

**Comment by SerialVelocity** [comment]
@leventov I haven't submitted it as a bug yet, I may do in a couple of days when I have some more time, but here's a repro (writing from memory):
```java
import java.util.function.BiFunction;

public class Bob {
  private static int getVarArgsSize(Bob... bobs) {
    return bobs.length;
  }

  private static int runBiFunction(BiFunction<Bob, Bob, Integer> function, Bob a, Bob b) {
    return function.apply(a, b);
  }

  public static void main(String[] args) {
    assert runBiFunction(Bob::getVarArgsSize, new Bob(), new Bob()) == 2;
  }
}
```
